atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
+        """ Create a command to lock slot for pipelinerun on remote host """
+        filepath = os.path.join(RELATIVE_PATH, file_name)
+        _cmd = (f'if grep -qw {{}} {filepath} ; then echo {quote(payload)} > '
+                f'{filepath} ;else exit 1; fi') #  TODO: Test with remote host that this is actually working with Paramiko
+        return f'flock {filepath} -c {quote(_cmd)}'
+
+    def lock(self, slot_id: int, prid: str):
+        logger.debug("Locking slot %s on host %s with pipelinerun ID %s",
+                     slot, self._hostname, prid)
+
+        payload = json.dumps({'prid': prid, 'locked': str(datetime.now())})
+
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_write_slot_in_progress(f'slot_{slot_id}.json', payload)
+            _, _, exitcode = self._ssh_run_remote_cmd(client, cmd)
+
+            if exitcode != 0:
+                raise LockError(f'Attempt to lock slot {slot_id} with pipelinerun ID {prid} '
+                                'returned non-zero exitcode')
+
+        logger.debug("%s - slot %s locked with pipelinerun ID %s", self._hostname, slot_id, prid)
+
+    def unlock(self, slot_id: int, prid: int):
+        logger.debug("Unlocking slot %s on host %s with pipelinerun ID %s",
+                     slot_id, self._hostname, prid)
+
+        with self._ssh_session() as client:
+            #  TODO: Is critical section needed here? -- Start
+            cmd = self.get_cmd_to_read_from_json(f'slot_{slot_id}.json')
+            slot = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+            if slot:
+                self._validate_resource_lock(slot)
+
+                if slot['prid'] != prid:
+                    logger.warning("%s - slot %s, Found pipelinerun ID: %s, expected %s. Unlocking"
+                                   " skipped!", self._hostname, slot_id, slot['prid'], prid)
+                    return
+
+                cmd = self.get_cmd_to_write_empty_slot(slot_id)
+                _, _, exitcode = self._ssh_run_remote_cmd(client, cmd)
+                #  TODO: Is critical section needed here? -- End (same for locking)
+
+                if exitcode != 0:
+                    logger.warning("Unlocking slot %s on host %s failed!", slot_id, self._hostname)
+            else:
+                logger.warning("%s - slot %s is available. Unlocking skipped!",
+                               resource_info['hostname'], slot_id)
+
+
+class RemoteHostsPool:
+
+    def __init__(self, hosts: List[RemoteHost]):
+        """
+        :param hosts: List[RemoteHost], List of Remote hosts
+        """
+        self.hosts = hosts
+
+    @classmethod
+    def from_config(cls, config: dict):
+        """ Instantiate remote hosts loaded from configmap
+
+        :param config: dict, Arch specific remote hosts dictionary from configmap
+
+        Example:
+
+        hostname-remote-host1:
+          enabled: true
```

Comment:
should this filter out not enabled hosts?

Reply:
depends if we want to do selection of enabled hosts here or on OSBS side. I am OK with selecting them here.

Reply:
Ok enabled hosts will be selected here in resource_manager.